[Dist Debugability] Refactor and fix DDP model check during init

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52966 [DDP Logging] Log comm. hook in ddp logging
* #52965 [DDP Logging] Log nccl_async_error_handling
* #52957 [Dist Debugality] Log key DDP metrics to stderr under debug mode.
* **#52887 [Dist Debugability] Refactor and fix DDP model check during init**

This diff changes the way to do model consistency check (i.e. `_verify_replicas_across_processes`) in DDP.

There were a few things that could be improved with the way we verify model across processes in DDP initialization:

1. We should do this check before syncing module states in DDP init, otherwise with Gloo backend this will throw but we would like to throw the error corresponding to different models on different ranks. To do this, we move the methods to be standalone C++ functions (not part of reducer) and move this check to before synchronizing parameters.
2. Refactor DDP init in the following ways:
- Run model consistency check before creating reducer, 2
- add helper functions to build params to pass into reducer
- add helper function to call `_verify_model_across_ranks`
- move `def parameters` to a helper function `_get_parameters` to be used more broadly within DDP

In follow up changes we will add the ability to detect which rank had inconsistent model (https://github.com/pytorch/pytorch/issues/52876 would be useful for this to determine which ranks(s) had errors).

Differential Revision: [D26565290](https://our.internmc.facebook.com/intern/diff/D26565290/)

rohan-varma committed:
Refactor/fix DDP model check during init

This diff changes the way to do model consistency check (i.e. `_verify_replicas_across_processes`) in DDP.

There were a few things that could be improved with the way we verify model across processes in DDP initialization:

1. We should do this check before syncing module states in DDP init, otherwise with Gloo backend this will throw but we would like to throw the error corresponding to different models on different ranks. To do this, we move the methods to be standalone C++ functions (not part of reducer) and move this check to before synchronizing parameters.
2. Refactor DDP init in the following ways:
- Run model consistency check before creating reducer, 2
- add helper functions to build params to pass into reducer
- add helper function to call `_verify_model_across_ranks`
- move `def parameters` to a helper function `_get_parameters` to be used more broadly within DDP

In follow up changes we will add the ability to detect which rank had inconsistent model (https://github.com/pytorch/pytorch/issues/52876 would be useful for this to determine which ranks(s) had errors).

Differential Revision: [D26565290](https://our.internmc.facebook.com/intern/diff/D26565290/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Debugability] Refactor and fix DDP model check during init"


This diff changes the way to do model consistency check (i.e. `_verify_replicas_across_processes`) in DDP.

There were a few things that could be improved with the way we verify model across processes in DDP initialization:

1. We should do this check before syncing module states in DDP init, otherwise with Gloo backend this will throw but we would like to throw the error corresponding to different models on different ranks. To do this, we move the methods to be standalone C++ functions (not part of reducer) and move this check to before synchronizing parameters.
2. Refactor DDP init in the following ways:
- Run model consistency check before creating reducer, 2
- add helper functions to build params to pass into reducer
- add helper function to call `_verify_model_across_ranks`
- move `def parameters` to a helper function `_get_parameters` to be used more broadly within DDP

In follow up changes we will add the ability to detect which rank had inconsistent model (https://github.com/pytorch/pytorch/issues/52876 would be useful for this to determine which ranks(s) had errors).

Differential Revision: [D26565290](https://our.internmc.facebook.com/intern/diff/D26565290/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Debugability] Refactor and fix DDP model check during init"


This diff changes the way to do model consistency check (i.e. `_verify_replicas_across_processes`) in DDP.

There were a few things that could be improved with the way we verify model across processes in DDP initialization:

1. We should do this check before syncing module states in DDP init, otherwise with Gloo backend this will throw but we would like to throw the error corresponding to different models on different ranks. To do this, we move the methods to be standalone C++ functions (not part of reducer) and move this check to before synchronizing parameters.
2. Refactor DDP init in the following ways:
- Run model consistency check before creating reducer, 2
- add helper functions to build params to pass into reducer
- add helper function to call `_verify_model_across_ranks`
- move `def parameters` to a helper function `_get_parameters` to be used more broadly within DDP

In follow up changes we will add the ability to detect which rank had inconsistent model (https://github.com/pytorch/pytorch/issues/52876 would be useful for this to determine which ranks(s) had errors).

Differential Revision: [D26565290](https://our.internmc.facebook.com/intern/diff/D26565290/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Debugability] Refactor and fix DDP model check during init"


This diff changes the way to do model consistency check (i.e. `_verify_replicas_across_processes`) in DDP.

There were a few things that could be improved with the way we verify model across processes in DDP initialization:

1. We should do this check before syncing module states in DDP init, otherwise with Gloo backend this will throw but we would like to throw the error corresponding to different models on different ranks. To do this, we move the methods to be standalone C++ functions (not part of reducer) and move this check to before synchronizing parameters.
2. Refactor DDP init in the following ways:
- Run model consistency check before creating reducer, 2
- add helper functions to build params to pass into reducer
- add helper function to call `_verify_model_across_ranks`
- move `def parameters` to a helper function `_get_parameters` to be used more broadly within DDP

In follow up changes we will add the ability to detect which rank had inconsistent model (https://github.com/pytorch/pytorch/issues/52876 would be useful for this to determine which ranks(s) had errors).

Differential Revision: [D26565290](https://our.internmc.facebook.com/intern/diff/D26565290/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -442,10 +442,24 @@ def __init__(self, module, device_ids=None,
+        # see Note: "Gradient Layout Contract" in Reducer::initialize_buckets).
+        if self.device_ids is not None and len(self.device_ids) > 1:
+            dist._verify_replicas_within_process(parameters, expect_sparse_gradient)
+        self._verify_model_across_ranks(parameters)
```
nit: directly call dist._verify_replica0_across_processes(self.process_group, parameters)?

rohan-varma replied:
The motivation for helper function was that after this diff, we will add CPU based barrier (https://github.com/pytorch/pytorch/issues/52876)  in debug mode, to report which ranks had an inconsistent model to help with a debug (and for all processes to throw a deterministic error that can be handled)

rohan-varma replied:
Changed it to directly call the cpp function for now, for the above comment, we can address it in that diff itself.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -442,10 +442,24 @@ def __init__(self, module, device_ids=None,
+            dist._verify_replicas_within_process(parameters, expect_sparse_gradient)
+        self._verify_model_across_ranks(parameters)
+        # Sync params and buffers. Ensures all DDP models start off at the same value.
         self._sync_params_and_buffers(authoritative_rank=0)
```
possibly we should still _sync_params_and_buffers first and then verify model consistency. it is possible that params are not exactly the same among ranks at the beginning, and this sync can make sure params are the same before verifying

rohan-varma replied:
@zhaojuanmao Syncing params and buffers before verifying model replicas was the root cause of why this didn't work with Gloo before, though. In particular, `_sync_params_and_buffers` will ensure all params have the same starting values, but if parameter shapes are different, at least `gloo` will throw an error (not sure why NCCL gets past this, but might be due to the async nature of the collective).

On the other hand, `_verify_model_across_ranks` doesn't actually check the _values_ of the parameters, only that they have the same shape. So even if they have different values, that is fine, as we are only interested in ensuring all params have the same shape so that there's no hang during training, and the values will be filled in consistently after we do the shape check.

zhaojuanmao replied:
I see, that makes sense!

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -459,31 +473,84 @@ def _sync_params_and_buffers(self, authoritative_rank=0):
+        (5) Logging constructin-time DDP logging data
+        (6) passing a handle of DDP to SyncBatchNorm Layer
         """
+        self.modules_params = [list(self._get_parameters(m)) for m in self._module_copies]
+        # Collect buffers for modules, filtering out buffers that should be ignored.
+        named_module_buffers = [
+            [(buffer, buffer_name) for buffer_name, buffer in m.named_buffers()]
+            for m in self._module_copies
+        ]
+        self.modules_buffers = [
+            [
+                buffer
+                for (buffer, buffer_name) in module_buffers
+                if buffer_name not in self.parameters_to_ignore
+            ]
+            for module_buffers in named_module_buffers
+        ]
```
nit: feel it can be moved into _build_params_for_reducer() as well

zhaojuanmao commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1287,6 +1287,20 @@ that adds a prefix to each key inserted to the store.
+      py::call_guard<py::gil_scoped_release>());
+
+  module.def(
+      "_verify_replica0_across_processes",
```
nit: it can be renamed as "_verify_model_across_ranks", so it does not need to define a new function in python again

rohan-varma committed:
Update on "[Dist Debugability] Refactor and fix DDP model check during init"


This diff changes the way to do model consistency check (i.e. `_verify_replicas_across_processes`) in DDP.

There were a few things that could be improved with the way we verify model across processes in DDP initialization:

1. We should do this check before syncing module states in DDP init, otherwise with Gloo backend this will throw but we would like to throw the error corresponding to different models on different ranks. To do this, we move the methods to be standalone C++ functions (not part of reducer) and move this check to before synchronizing parameters.
2. Refactor DDP init in the following ways:
- Run model consistency check before creating reducer, 2
- add helper functions to build params to pass into reducer
- add helper function to call `_verify_model_across_ranks`
- move `def parameters` to a helper function `_get_parameters` to be used more broadly within DDP

In follow up changes we will add the ability to detect which rank had inconsistent model (https://github.com/pytorch/pytorch/issues/52876 would be useful for this to determine which ranks(s) had errors).

Differential Revision: [D26565290](https://our.internmc.facebook.com/intern/diff/D26565290/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Debugability] Refactor and fix DDP model check during init"


This diff changes the way to do model consistency check (i.e. `_verify_replicas_across_processes`) in DDP.

There were a few things that could be improved with the way we verify model across processes in DDP initialization:

1. We should do this check before syncing module states in DDP init, otherwise with Gloo backend this will throw but we would like to throw the error corresponding to different models on different ranks. To do this, we move the methods to be standalone C++ functions (not part of reducer) and move this check to before synchronizing parameters.
2. Refactor DDP init in the following ways:
- Run model consistency check before creating reducer, 2
- add helper functions to build params to pass into reducer
- add helper function to call `_verify_model_across_ranks`
- move `def parameters` to a helper function `_get_parameters` to be used more broadly within DDP

In follow up changes we will add the ability to detect which rank had inconsistent model (https://github.com/pytorch/pytorch/issues/52876 would be useful for this to determine which ranks(s) had errors).

Differential Revision: [D26565290](https://our.internmc.facebook.com/intern/diff/D26565290/)

[ghstack-poisoned]
